pinot-broker/src/main/java/org/apache/pinot/broker/api/AccessControl.java

```diff
@@ -42,21 +46,58 @@ default boolean hasAccess(RequesterIdentity requesterIdentity) {
 
   /**
    * Fine-grained access control on parsed broker request. May check table, column, permissions, etc.
-   *
+   * The default implementation is kept to have backward compatibility with the existing implementations
    * @param requesterIdentity requester identity
    * @param brokerRequest broker request (incl query)
    *
    * @return {@code true} if authorized, {@code false} otherwise
    */
-  boolean hasAccess(RequesterIdentity requesterIdentity, BrokerRequest brokerRequest);
+
+  default boolean hasAccess(RequesterIdentity requesterIdentity, BrokerRequest brokerRequest) {
+    return true;
```

Comment:
The high level idea is to only override one of `hasAccess` and `verifyAccess`, and both of them can be used. So I think we should `return verifyAccess(requesterIdentity, brokerRequest).hasAccess()`

Reply:
I thought a bit about this before arriving at an option to have the default set for both hasAccess() & verifyAccess() ( now authorize() ).


Now the baseQueryHandler is calling verifyAccess(), and the default implementation of verifyAccess() now calls hasAccess() - This is a safe case of all existing implementations of AccessControl to work  ( added a UT for this specific scenario ).

about having default for hasAccess() - The goal was to move away from hasAccess() to VerifyAccess() for all implementations . So as not to cause any friction during this intermediary phase, I added a default implementation. 

I'll proceed to mark hasAccess() as Deprecated. LMK your thoughts?

Reply:
+1 on marking it `@Deprecated`

Making it return `true` by default won't cause problem within Pinot, but my concern is that if there are client code (outside of Pinot repo) still calling `hasAccess()`, that one will break because all the existing implementation will return `true`. That is why it is safer to make them call each other so that as long as one of them is overridden, both of them can work properly

Reply:
> but my concern is that if there are client code (outside of Pinot repo) still calling `hasAccess()`, that one will break because all the existing implementation will return `true`

I don't think this would happen .

**Reason :** 
We are making both implementations of hasAccess() default only with this change . So by that all existing implementations would already have `hasAccess()` overridden . 
Only chance of getting `true` is any new implementation of `AccessControl` missing to override both `hasAccess()` and `authorize()` ( previously `verifyAccess()` ) . This chance of miss is again only going to in the transition phase till we get `hasAccess()` removed and make `authorize()` non-default hopefully in upcoming major pinot version future releases.

Please correct me if I misunderstood the scenario you are trying to say 

**Further thoughts :**
My concern with making them call each other is ,  any new implementation of `AccessControl` ( outside of pinot repo & by mistake not having `hasAccess()` or `authorize()` implemented ) and client calling either of `hasAccess()` or `authorize()` of that undesirable implementation will cause `StackOverflowError` .

As its currently marked as `@Deprecated` , the compiler would throw warning . 
Further more if any clients calling implementations of `hasAccess()` would work normally . 

To make it more rigid , instead of returning `true` from `hasAccess()` default implementation , we can probably throw an `DeprecationException()` / `NotImplementedException()` [ I prefer the former exception ]  .  LMK if this would work @Jackie-Jiang  ?

Reply:
I just realized that `BasicAuthAccessControlFactory` and `ZkBasicAuthAccessControlFactory` override both the methods, thus the problem I described won't happen.
I like the idea of making the default `hasAccess()` throw `UnsupportedOperationException` (we usually use this to represent unimplemented method) which can force caller to switch to the new API if there are any. In that case we might also want to remove the override in `BasicAuthAccessControlFactory` and `ZkBasicAuthAccessControlFactory`

Reply:
Made the changes @Jackie-Jiang  . PTAL .

> In that case we might also want to remove the override in BasicAuthAccessControlFactory and ZkBasicAuthAccessControlFactory

I hope you meant to not have overrides for hasAccess() in `BasicAuthAccessControlFactory` and `ZkBasicAuthAccessControlFactory` . If Yes , they have been already removed . Only overrides of `authorize()` is present there

_On a related note to this discussions :_ 
I left this as it is ,
```
default boolean hasAccess(RequesterIdentity requesterIdentity) {
    return true;
  }
```
and created a new equivalent `authorize(RequesterIdentity requesterIdentity)` which in default calls `hasAccess(requesterIdentity)` . I don't suspect any backward incompatibility problems here .